Improve packaging of individual functions

davidvanleeuwen opened the pull request:
This PR fixes a few things that are relevant to one of our projects. Please let me know what you think of this PR. After your go/no-go I'll check out the unit tests, because they aren't working yet (**don't merge this PR**).

* Instead of throwing all configs in `webpack`, I'd prefer to do it one by one sequentially. In projects with multiple functions, memory usage can become a big problem. Therefor I also added `cache: false` to it and don't do anything with the result.
* It runs through the functions and logs it to `serverless` (per function).
* Because the source is webpacked, files can become pretty big. Therefor I added the option to exclude everything, except for the function file itself.
* I also prefer if everything packaged keeps its original path, especially in bigger projects filenames can collide.
* In our case we also use the plugin `CopyWebpackPlugin` which moves over some modules that use c/c++. Therefore `node_modules` is excluded. I know, it's kind of a specific scenario, but I think other people might want to do this too for backends while all other files are included through webpack.

Commit:
Webpack functions one by one and don’t cache

Commit:
Keep folder structure and handler and only package that file

Commit:
Refactor

jogold commented:
Hello @davidvanleeuwen,

Thanks for the PR, I will have a look into it.

A few quick questions :
* I'm already excluding everything at the package level (https://github.com/goldwasserexchange/serverless-plugin-webpack/blob/master/src/index.js#L47) and including only the function at the function level (https://github.com/goldwasserexchange/serverless-plugin-webpack/blob/master/src/index.js#L53). You changed this because you wanted to add `node_modules/*` ?
* When using `Copy Webpack Plugin`, your copied node modules are then included in all zip files even if they are not used by all functions ?
* Wouldn't it be better to take function specific includes into account ?

jogold commented on src/lib/wpack.js:
```diff
@@ -57,22 +57,30 @@ const createConfigs = (fns, config, servicePath, defaultOutput, folder) =>
+            hash: false,
+            chunks: false,
+            version: false,
+            cache: false,
```
You meant `cached`? Also, this has no effect on caching, it only hides information about cached modules in the output. See https://webpack.js.org/configuration/stats/.

davidvanleeuwen replied:
Ye, you're right. Just checked everything off. The fix is not returning the resulted stats from the promise. Otherwise it stacks it and it'll run into a node memory error if it becomes too big.

davidvanleeuwen commented:
Thanks for your reply!

* It should be `exclude: ['**']` because I changed it to have subdirectories too. Changed that in your code.
* Correct, I added copying to `.webpack/node_modules` - which do get included in the artifact.
* Yes, but that's a choice - I guess it's hard to make it really function specific, other than making a webpack config per function.
* How would you suggest doing it?

Commit:
Fix excluding files

Commit:
Merge branch 'remote-master'

# Conflicts:
#	src/lib/functions.js

Commit:
Unneeded

jogold commented:
Hi @davidvanleeuwen, 

Have you seen the latest changes with excludes/includes ?

Do you still want to run webpack configs in series ? How many functions do you have in your projects ?

davidvanleeuwen commented:
Sorry, didn't have the change yet to look at it. Have been running into other things that need my attention. I'll try to look at it this week though and make a separate PR for running webpack in series (as configurable option in serverless.yml).

We have lots of functions actually which represent API endpoints.

davidvanleeuwen commented:
Using `include` is not really working for me, I should probably try it in a simple setup - because maybe some other things aren't working as expected.

jogold commented:
Closing this for now. Feel free to re-open.
